---
title: "community_analysis_prescribed-fires"
author: "T. Griffin"
date: "2024-04-23"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
##Goals
1. Load in phyloseq data with rooted tree.
2. Evaluate sequencing depth and remove sample.
3. Normalize the read counts between samples.
4. Calculate community dissimilarities. Numbers between 0 and 1. If 0, completely similar versus if they are 1, then theyâ€™re completely dissimilar.
  Sorensen: Shared Species as a binary value: Abundance-unweighted
  Bray-Curtis: Shared Abundant species: Abundance-weighted
  (Abundance-)Weighted UNIFRAC: Consider Abundant Species and where they f   all on the tree
5. Visualize the community data with two unconstrained Ordinations:
  PCoA: Linear Method. Eigenvalue = how much variation is explained by each     axis. Choose to view axis 1, 2, 3, etc. and plot them together.
  NMDS: Non-linear. Smush multiple Dimensions into 2 or 3 axes. Need to        report Stress value (ideally <0.15).
6. Run statistics with PERMANOVA and betadispR.


##Seed setting
```{r}
#this will set my seed again
set.seed(02262000)
```

##Load libraries
```{r}
#load libraries adn colors
pacman::p_load(tidyverse, devtools, phyloseq, patchwork, vegan,
               install = FALSE)

horizon_colors <- c(
  "A horizon" = "pink",
  "B horizon" = "green",
  "E horizon" = "orange")

#colors for burn year
by_colors <- c(
  "CH" = "blue",
  "1" = "purple",
  "2" = "brown",
  "4" = "yellow")
```

##Load Data
```{r}
#load the rooted tree
load("./phylo_tree/phytree_preprocessed_physeq.RData")
merged_phytree
```

##look at the read counts
```{r}
# Calculate the total number of reads per sample. This will tell me how many reads were per sample and I can see any outliers
raw_TotalSeqs_df  <- 
  merged_phytree %>%
  # calculate the sample read sums 
  sample_sums() %>%
  data.frame()
# name the column 
colnames(raw_TotalSeqs_df)[1] <- "TotalSeqs"
  
head(raw_TotalSeqs_df)
#low: 10470-SRR19625128, high: 148864-SRR19625131

# make histogram of raw reads 
raw_TotalSeqs_df %>%
  ggplot(aes(x = TotalSeqs)) + 
  geom_histogram(bins = 50) + 
  scale_x_continuous(limits = c(0, 160000)) + 
  labs(title = "Raw Sequencing Depth Distribution") + 
  theme_classic()

#these did not appear strange during the phylogenetic tree generation, i do not think that i need to remove them

```

##normalize read counts
```{r}
#normalization makes comparisons between samples more accurate since different variables can affect read counts
matround <- function(x){trunc(x+0.5)}

scale_reads <- function(physeq, n = min(sample_sums(physeq)), round = "round") {
  
  # transform counts to n
  physeq.scale <- transform_sample_counts(physeq, function(x) {(n * x/sum(x))})
  
  # Pick the rounding functions
  if (round == "floor"){
    otu_table(physeq.scale) <- floor(otu_table(physeq.scale))
  } else if (round == "round"){
    otu_table(physeq.scale) <- round(otu_table(physeq.scale))
  } else if (round == "matround"){
    otu_table(physeq.scale) <- matround(otu_table(physeq.scale))
  }
  
  # Prune taxa and return new phyloseq object
  physeq.scale <- prune_taxa(taxa_sums(physeq.scale) > 0, physeq.scale)
  return(physeq.scale)
}

min(sample_sums(merged_phytree))
```

##Scale reads and check distribution
```{r}
scaled_merged_phytree <- 
  merged_phytree %>%
  scale_reads(round = "matround")

# Calculate the read depth 
scaled_TotalSeqs_df <- 
  scaled_merged_phytree %>%
  sample_sums() %>%
  data.frame()
colnames(scaled_TotalSeqs_df)[1] <-"TotalSeqs"

# Inspect
head(scaled_TotalSeqs_df)
max(scaled_TotalSeqs_df) - min(scaled_TotalSeqs_df)

scaled_TotalSeqs_df %>%
  ggplot(aes(x = TotalSeqs)) + 
  geom_histogram(bins = 50) + 
  scale_x_continuous(limits = c(0, 160000)) + 
  labs(title = "Scaled Sequencing Depth at 10470") + 
  theme_classic()
#samples seem to be at similar range
```

##PCoA
```{r}
# Calculate sorensen dissimilarity: Abundance-unweighted of shared taxa
scaled_soren_pcoa <-  
  ordinate(
    physeq = scaled_merged_phytree,
    method = "PCoA",
    distance = "bray", binary = TRUE)

#str(scaled_soren_pcoa)

# Plot the ordination 
soren_horizon_pcoa <- plot_ordination(
  physeq = scaled_merged_phytree,
  ordination = scaled_soren_pcoa,
  color = "horizon",
  shape = "horizon",
  title = "Sorensen Horizon PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = horizon)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = horizon_colors) + 
  theme_bw()
# Show the plot 
soren_horizon_pcoa

#species found in individual soil horizons seem to cluster together by sorensen

# Calculate the BC distance
scaled_BC_pcoa <- 
  ordinate(
    physeq = scaled_merged_phytree,
    method = "PCoA",
    distance = "bray")

# Plot the PCoA
bray_horizon_pcoa <- 
  plot_ordination(
    physeq = scaled_merged_phytree,
    ordination = scaled_BC_pcoa,
    color = "horizon",
    shape = "horizon",
    title = "Bray-Curtis Horizon PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = horizon)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = c(horizon_colors)) + 
  theme_bw()
bray_horizon_pcoa

#similar results for the bc; similar clustering based on soil horizon

unfiltered_unifrac <-
  ordinate(
    physeq=scaled_merged_phytree,
    method="PCoA",
    distance="wunifrac")

w_unifrac <- plot_ordination(
  physeq=scaled_merged_phytree,
  ordination=unfiltered_unifrac,
  color="horizon",
  title= "Weighted Horizon Unifrac PCoA")+
  geom_point(size=4, alpha=0.6, aes(color= horizon))+
  scale_color_manual(values=horizon_colors)
w_unifrac

#combine the plots
(soren_horizon_pcoa + theme(legend.position = "none")) + 
  (bray_horizon_pcoa + theme(legend.position = "none")) + 
    (w_unifrac + theme(legend.position = "none"))

```

##NMDS
```{r}
scaled_wUNI_nmds <- 
  ordinate(
    physeq = scaled_merged_phytree,
    method = "NMDS",
    distance = "wunifrac")

wUNI_station_nmds <- 
  plot_ordination(
    physeq = scaled_merged_phytree,
    ordination = scaled_wUNI_nmds,
    color = "horizon",
    shape = "horizon",
    title = "Weighted Unifrac Horizon NMDS") +
  geom_point(size=5, alpha = 0.5, aes(color = horizon)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = c(horizon_colors)) + 
  theme_bw()
wUNI_station_nmds

```

##PERMANOVA
```{r}
# Calculate all three of the distance matrices
scaled_sorensen_dist <- phyloseq::distance(scaled_merged_phytree, method = "bray", binary = TRUE)
scaled_bray_dist <- phyloseq::distance(scaled_merged_phytree, method = "bray")
scaled_wUnifrac_dist <- phyloseq::distance(scaled_merged_phytree, method = "wunifrac")

# make a data frame from the sample_data
# All distance matrices will be the same metadata because they 
# originate from the same phyloseq object. 
metadata_horizon <- data.frame(sample_data(scaled_merged_phytree))

# Adonis test
# In this example we are testing the hypothesis that the five stations
# that were collected have different centroids in the ordination space 
# for each of the dissimilarity metrics, we are using a discrete variable 
adonis2(scaled_sorensen_dist ~ horizon, data = metadata_horizon)

adonis2(scaled_bray_dist ~ horizon, data=metadata_horizon)

adonis2(scaled_wUnifrac_dist ~ horizon, data=metadata_horizon)

```

