---
title: "Phyloseq_amplicon-proj_runthrough"
author: "T. Griffin"
date: "2024-03-18"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
#send this phyloseq object to this folder

knitr::opts_chunk$set(echo = TRUE, 
                      fig.align = "center",
                      fig.path = "./preprocessing_phyloseq")
```

## loading libraries
```{r loading-libraries}
library(devtools)
library(phyloseq)
library(tidyverse)
```
## Goals

Here, we will  process the data into a phyloseq object
We will need:
1. ASV table
2. Taxonomy table
3. Track Reads (metadata)

Then, we will remove the following:

1. Remove mitochondria (BECAUSE THEY HAVE THEIR OWN 16S DNA)
2. Remove chloroplasts
3. Remove samples without "enough" readds

Finally, we will write a data file of phyloseq output.

## load the ASV count table and tax table
```{r load-the-ASV-count-table-and-tax-table}
<<<<<<< HEAD
load("/local/workdir/tag92/git_repos/comp_micro_proj/Data/01_DADA2/output_DADA2_ASV_counts.RData")
=======
load("./DADA2/Data/output_DADA2_ASV_counts.RData")
>>>>>>> 5fb24fe06b44024b1e979f22b3b3ede6b0b8ab55

# Inspect asv_tab
head(asv_tab)[,1:5]

sample_names <- colnames(asv_tab)
samples_fixed <- sapply(strsplit(basename(sample_names), "_"), `[`,1)
head(samples_fixed)

colnames(asv_tab) <- samples_fixed 
str(asv_tab)
#this will fix the names in the table
```
```{r load-taxonomy-table}
<<<<<<< HEAD
tax_df <- read.table("/local/workdir/tag92/git_repos/comp_micro_proj/Data/01_DADA2/output_DADA2_ASV_taxonomy.tsv", sep = "\t", skip = 1)
=======
tax_df <- read.table("./DADA2/Data/output_DADA2_ASV_taxonomy.tsv", sep = "\t", skip = 1)
>>>>>>> 5fb24fe06b44024b1e979f22b3b3ede6b0b8ab55

head(tax_df)

#fix column names
colnames(tax_df) <- c("ASV", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species", "ASV", "ASVseq")

head(tax_df)

#Taxonomy table matrix
tax_mat <- 
  tax_df %>%
  tibble::column_to_rownames(., var = "ASV") %>%
  as.matrix()
```

## Load tracked reads
```{r load-tracked-reads}
<<<<<<< HEAD
load("/local/workdir/tag92/git_repos/comp_micro_proj/Data/01_DADA2/output_DADA2_track_read_counts.RData")
=======
load("./DADA2/Data/output_DADA2_track_read_counts.RData")
>>>>>>> 5fb24fe06b44024b1e979f22b3b3ede6b0b8ab55
#Take a look at the data
track_counts_df

head(track_counts_df)
dim(track_counts_df)

#load in metadata
<<<<<<< HEAD
metadata_df <- read.csv("/local/workdir/tag92/git_repos/comp_micro_proj/metadata/comp_micro_proj_metadata.csv")
=======
metadata_df <- read.csv("./metadata/comp_micro_proj_metadata.csv")
>>>>>>> 5fb24fe06b44024b1e979f22b3b3ede6b0b8ab55
#since the metadata does not separate the burn year into their own column, I have to do that myself. I cannot do this in one step, so I will create two dataframes: one that contains all values for the numerical years, and another with the values for the controls since the control burn years have characters. I am fixing this here so maybe I can use the burn year column in later steps

#this first one is for the rows with numbers
numcolumns_metadata_df <- tidyr::extract(metadata_df, Site, c('burn_year', 'Plot'), '(^\\d)(.*)', convert = FALSE)

#this is for the rows with characters
controlcolumns_metadata_df <- tidyr::extract(metadata_df, Site, c('burn_year', 'Plot'), '(CH)(.*)', convert = FALSE)

#now I need to merge them to get the full dataframe with all the correct values
burnyear_metadata_df <- rbind(
filter(controlcolumns_metadata_df, !is.na(Plot)),
filter(numcolumns_metadata_df, !is.na(Plot)))


view(burnyear_metadata_df)
dim(burnyear_metadata_df)
colnames(burnyear_metadata_df)

head(burnyear_metadata_df)
head(track_counts_df)

#the column "names" was not in the metadata file, so I changed this to be able to join the dataframes
colnames(burnyear_metadata_df)[1] <- "names"

metadata_track_reads_df <- burnyear_metadata_df %>%
  left_join(., track_counts_df, by = "names")

head(metadata_track_reads_df)

#the row names are just numbers right now, so I should change them to more meaningful sample names
row.names(metadata_track_reads_df)

#the actual change!
row.names(metadata_track_reads_df) <- metadata_track_reads_df$names

row.names(metadata_track_reads_df)

row.names(metadata_track_reads_df)

head(metadata_track_reads_df)

```

## Hand off to phyloseq
```{r hand-off-to-phyloseq}
dim(asv_tab)

dim(tax_mat)

# Intuition check (ask what exactly this does)
stopifnot(row.names(asv_tab) == row.names(tax_mat))

# Construct the phyloseq object 
raw_physeq <- phyloseq(otu_table(asv_tab, taxa_are_rows = TRUE),
                       sample_data(metadata_track_reads_df),
                       tax_table(tax_mat))

#saving
<<<<<<< HEAD
save(raw_physeq, file = "/local/workdir/tag92/git_repos/comp_micro_proj/Data/raw_physeq.RData")
=======
save(raw_physeq, file = "./preprocessing_phyloseq/raw_physeq.RData")
>>>>>>> 5fb24fe06b44024b1e979f22b3b3ede6b0b8ab55
```

## Remove mitochondrias and chloroplasts
```{r remove-mitochondrias-and-chloroplasts}
#make new physeq without mito. and chloro.

noChloros_physeq <- raw_physeq %>%
  subset_taxa(Order != "Chloroplast" | is.na(Order))
  
#How many taxa were chloroplasts?
num_chloros_ASVs <- ntaxa(raw_physeq) - ntaxa(noChloros_physeq)
num_chloros_ASVs

noChloros_physeq %>%
tax_table() %>%
data.frame() %>%
  view()

#remove mitochondria
nochlorosmitos_physeq <- noChloros_physeq %>%
  subset_taxa(Family != "Mitochondria" | is.na(Family))

#How many mitochondrial ASVs?
num_mito_ASVs <- ntaxa(noChloros_physeq) - ntaxa(nochlorosmitos_physeq)
num_mito_ASVs

#How many total asvs were removed>
ntaxa(raw_physeq) - ntaxa(nochlorosmitos_physeq)

#what proportion of ASVs were kept?
ntaxa(nochlorosmitos_physeq)/ntaxa(raw_physeq)

#view(nochlorosmitos_physeq)

```

```{r removing-controls}
#I don't think our controls were included in the dataset

```

## Evaluate sequencing depth
```{r evaluate-sequencing-depth}
#the current object is nochlorosmitos_physeq
seqSums_df <- 
  nochlorosmitos_physeq %>%
  otu_table() %>%
  # Sum each sample column 
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "names") %>%
  left_join(., metadata_track_reads_df, by = "names") 

# Rename second column 
colnames(seqSums_df)[2] <- "TotalSeqs"

# check
dim(seqSums_df)

#show number of seqs for each sample
seqSums_df %>%
  dplyr::select(names, TotalSeqs) %>%
  arrange(TotalSeqs) %>%
  head()

#bar plot
seqSums_df %>%
  ggplot(aes(x=reorder(names, TotalSeqs), y = TotalSeqs,
             fill = burn_year)) + 
  geom_bar(stat = "identity") 

#density plot
seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = burn_year)) +
  geom_density(alpha = 0.5)

#density plot for horizon
seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = horizon)) +
  geom_density(alpha = 0.5)
```
<<<<<<< HEAD
Burn year four has higher depth than the other burn years and controls. For horizon, sequencing depth does not vary by much. 

=======
>>>>>>> 5fb24fe06b44024b1e979f22b3b3ede6b0b8ab55

## remove lowly reads
```{r remove-lowly-reads}
min(sample_sums(nochlorosmitos_physeq))
#10464

max(sample_sums(nochlorosmitos_physeq))
#148519

#change the name of the object to match subsequent analyses since I have rerun this one to include the burn year columns

NEW_raw_physeq <- nochlorosmitos_physeq
```

<<<<<<< HEAD
From the density plots, a lot of reads are less than 50000, and only a few exceed 100000. This does not appear problematic. 
=======
#From the density plots, a lot of reads are less than 50000, and only a few exceed 100000. This does not appear problematic. 
>>>>>>> 5fb24fe06b44024b1e979f22b3b3ede6b0b8ab55


## Save this phyloseq object
```{r save-this-phyloseq-object}
save(NEW_raw_physeq,
<<<<<<< HEAD
     file = "/local/workdir/tag92/git_repos/comp_micro_proj/Data/NEW_raw_physeq.RData")
=======
     file = "./preprocessing_phyloseq/NEW_raw_physeq.RData")
>>>>>>> 5fb24fe06b44024b1e979f22b3b3ede6b0b8ab55
```

# Session info
```{r session-info}
devtools::session_info()