---
title: "05_communityanalysis"
author: "Ri Desroches"
date: "`r Sys.Date()`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "../figures/05_communityanalysis/")

```

# Setup 

## set seed
```{r set-seed}

set.seed(0110)

```

## Load Packages

```{r load-packages}
pacman::p_load(tidyverse, devtools, phyloseq, patchwork, vegan,
               install = FALSE)
```


## Load data

```{r load-data}
load("data/04_phylogeny/phytree_preprocessed_physeq.RData")

#set color schemes as well

#color vectors based on burn time frame
timeframe_colors <- c(
  "1"="darkorange1",
  "2"="goldenrod2",
  "4"="olivedrab",
  "unburned"="darkgreen")


#color vectors based on sample horizon
sample_horizon_colors <- c(
  "A horizon"="rosybrown1",
  "B horizon"="lightsalmon3",
  "E horizon"= "darkred")
```


## Goals :

1. Evaluate sequencing depth
2. Remove low-depth samples
3. Normalize read counts
4. Calculate community dissimilarities with three metrics:
  a. Sorensen
  b. Bray-Curtis
  c. Abundance weighted unifrac
5. Visualize the community data with:
  a. PCoA
  b. NMDS
6. Stats analyses

# Unfiltered dataset

## Explore read counts

```{r explore-read-counts}
raw_totalseqs_df <-
  unfiltered_rooted_tree_physeq %>%
  sample_sums() %>%
  data.frame()

colnames(raw_totalseqs_df)[1] <- "TotalSeqs"

head(raw_totalseqs_df)


# make histogram

raw_totalseqs_df %>%
  ggplot(aes(x = TotalSeqs)) + 
  geom_histogram()+ 
  labs(title = "Raw Sequencing Depth Distribution")
```

The lowest seq depth sample is 14380 seqs

Keep all samples, there is a very wide range of seqs

The only outlier is the one with 200k+ sequences, see if it stands out during community analyses

That sample is : SRR19625131

```{r SRR19625131-sample-info}
sample_data <- unfiltered_rooted_tree_physeq@sam_data %>%
  tibble::as_tibble()

dplyr::filter(sample_data, names == "SRR19625131")

```

That sample is an A horizon, 1-year burned forest

I also see that I have to separate burn year into its own metadata column again, but wrote the code previously and it didn't save it as an object. Oops!

Fixing before moving on

## Fixing metadata.. again

```{r adding-burn-year-metadata}
#Used code from 03_biodiversity.Rmd


#Information to add:
  # burn status (1 year, 2 year, 4 year, unburned)
  #separate from "sample site"
  # goal is to have a column that just has timeframe of burn without A-F sampling information


#extract burn time frame information from one of the 4 objects
unrooted_physeq@sam_data$Site

burn_timeframe <- gsub("^CH", "unburned", gsub("[A-Z]$", "", unrooted_physeq@sam_data$Site))

# fix unrooted physeq

unrooted_physeq@sam_data$Site <- burn_timeframe

sample_data(unrooted_physeq)$burn_timeframe <-
  sample_data(unrooted_physeq)$Site

# fix midroot physeq

unfiltered_rooted_tree_physeq@sam_data$Site <- burn_timeframe

sample_data(unfiltered_rooted_tree_physeq)$burn_timeframe <- 
  sample_data(unfiltered_rooted_tree_physeq)$Site

# fix unrooted filtered physeq

unrooted_physeq_filtered@sam_data$Site <- burn_timeframe

sample_data(unrooted_physeq_filtered)$burn_timeframe <-
  sample_data(unrooted_physeq_filtered)$Site

# fix rooted filtered physeq

filtered_rooted_tree_physeq@sam_data$Site <- burn_timeframe

sample_data(filtered_rooted_tree_physeq)$burn_timeframe <-
  sample_data(filtered_rooted_tree_physeq)$Site

```

## Normalize read counts - import function

```{r normalize-read-counts-function}

# Function to scale reads: http://deneflab.github.io/MicrobeMiseq/ 
# Scales reads by 
# 1) taking proportions
# 2) multiplying by a given library size of n
# 3) rounding 
# Default for n is the minimum sample size in your library
# Default for round is floor

matround <- function(x){trunc(x+0.5)}

scale_reads <- function(physeq, n = min(sample_sums(physeq)), round = "round") {
  
  # transform counts to n
  physeq.scale <- transform_sample_counts(physeq, function(x) {(n * x/sum(x))})
  
  # Pick the rounding functions
  if (round == "floor"){
    otu_table(physeq.scale) <- floor(otu_table(physeq.scale))
  } else if (round == "round"){
    otu_table(physeq.scale) <- round(otu_table(physeq.scale))
  } else if (round == "matround"){
    otu_table(physeq.scale) <- matround(otu_table(physeq.scale))
  }
  
  # Prune taxa and return new phyloseq object
  physeq.scale <- prune_taxa(taxa_sums(physeq.scale) > 0, physeq.scale)
  return(physeq.scale)
}

```

## Normalize read counts 

```{r normalize-read-counts}
min(sample_sums(unfiltered_rooted_tree_physeq))

# scale reads to 14380

scaled_unfiltered_rooted_physeq <-
  unfiltered_rooted_tree_physeq %>%
  scale_reads(round="matround")


scaled_totalseqs_df <-
  scaled_unfiltered_rooted_physeq %>%
  sample_sums() %>%
  data.frame() 

colnames(scaled_totalseqs_df) <- "seq_counts"

head(scaled_totalseqs_df)

max(scaled_totalseqs_df) - min(scaled_totalseqs_df)

#Plot as histogram

scaled_totalseqs_df %>%
  ggplot(aes(x=seq_counts)) +
  geom_histogram()

```

Looks good, they are all in a similar range.

## Community analyses

### Sorensen

```{r Sorensen}

unfiltered_sorensen <-
  ordinate(
    physeq=scaled_unfiltered_rooted_physeq,
    method="PCoA",
    distance="bray", binary=TRUE)

plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_sorensen,
  color="burn_timeframe",
  title="Unfiltered-Kingdom Dataset Sorensen PCoA")+
  geom_point(size=4, alpha=0.6, aes(color=burn_timeframe))+
  scale_color_manual(values=timeframe_colors)


plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_sorensen,
  color="horizon",
  title="Unfiltered-Kingdom Dataset Sorensen PCoA")+
  geom_point(size=4, alpha=0.6, aes(color=horizon))+
  scale_color_manual(values=sample_horizon_colors)

```

### Bray-Curtis

```{r bray-curtis}
unfiltered_bray <-
  ordinate(
    physeq=scaled_unfiltered_rooted_physeq,
    method="PCoA",
    distance="bray")

plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_bray,
  color="burn_timeframe",
  title="Unfiltered-Kingdom Dataset BC PCoA")+
  geom_point(size=4, alpha=0.6, aes(color=burn_timeframe))+
  scale_color_manual(values=timeframe_colors)


plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_bray,
  color="horizon",
  title="Unfiltered-Kingdom Dataset BC PCoA")+
  geom_point(size=4, alpha=0.6, aes(color=horizon))+
  scale_color_manual(values=sample_horizon_colors)

```

### Weighted Unifrac

```{r weighted-unifrac}

unfiltered_unifrac <-
  ordinate(
    physeq=scaled_unfiltered_rooted_physeq,
    method="PCoA",
    distance="wunifrac")

plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_unifrac,
  color="burn_timeframe",
  title="Unfiltered-Kingdom Dataset Unweighted Unifrac PCoA")+
  geom_point(size=4, alpha=0.6, aes(color=burn_timeframe))+
  scale_color_manual(values=timeframe_colors)


plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_unifrac,
  color="horizon",
  title="Unfiltered-Kingdom Dataset Weighted Unifrac PCoA")+
  geom_point(size=4, alpha=0.6, aes(color=horizon))+
  scale_color_manual(values=sample_horizon_colors)

```

### Permanova

```{r permanova}

scaled_sorensen <- phyloseq::distance(scaled_unfiltered_rooted_physeq, method = "bray", binary = TRUE)
scaled_bray <- phyloseq::distance(scaled_unfiltered_rooted_physeq, method = "bray")
scaled_unifrac <- phyloseq::distance(scaled_unfiltered_rooted_physeq, method = "wunifrac")

metadata <- data.frame(sample_data(scaled_unfiltered_rooted_physeq))

adonis2(scaled_sorensen ~ burn_timeframe, data=metadata)

adonis2(scaled_bray ~ burn_timeframe, data=metadata)

adonis2(scaled_unifrac ~ burn_timeframe, data=metadata)

adonis2(scaled_unifrac ~ horizon, data=metadata)

```


# Filtered dataset