---
title: "community_analysis_prescribed-fires"
author: "T. Griffin"
date: "2024-04-23"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Goals
1. Load in phyloseq data with rooted tree.
2. Evaluate sequencing depth and remove sample.
3. Normalize the read counts between samples.
4. Calculate community dissimilarities. Numbers between 0 and 1. If 0, completely similar versus if they are 1, then theyâ€™re completely dissimilar.
  Sorensen: Shared Species as a binary value: Abundance-unweighted
  Bray-Curtis: Shared Abundant species: Abundance-weighted
  (Abundance-)Weighted UNIFRAC: Consider Abundant Species and where they f   all on the tree
5. Visualize the community data with two unconstrained Ordinations:
  PCoA: Linear Method. Eigenvalue = how much variation is explained by each     axis. Choose to view axis 1, 2, 3, etc. and plot them together.
  NMDS: Non-linear. Smush multiple Dimensions into 2 or 3 axes. Need to        report Stress value (ideally <0.15).
6. Run statistics with PERMANOVA and betadispR.

I am hoping to see if burned soils causes some extremophilic bacteria to start appearing with more burning!

## Set Seed
```{r set-seed}
#this will set my seed again
set.seed(02262000)
```

## Load libraries
```{r load-libraries-and-colors}
#load libraries adn colors
library(tidyverse)
library(devtools) 
library(phyloseq) 
library(patchwork)
library(vegan)

horizon_colors <- c(
  "A horizon" = "pink",
  "B horizon" = "green",
  "E horizon" = "orange")

#colors for burn year
by_colors <- c(
  "CH" = "blue",
  "1" = "purple",
  "2" = "brown",
  "4" = "yellow")
```

## Load Data
```{r load-data}
#load the rooted tree
load("/local/workdir/tag92/git_repos/comp_micro_proj/Data/03_Phylogenetic_Tree/merged_phytree_preprocessed_physeq.RData")
merged_phytree
```

## Read counts
```{r read-counts}
# Calculate the total number of reads per sample. This will tell me how many reads were per sample and I can see any outliers
raw_TotalSeqs_df  <- 
  merged_phytree %>%
  # calculate the sample read sums 
  sample_sums() %>%
  data.frame()
# name the column 
colnames(raw_TotalSeqs_df)[1] <- "TotalSeqs"
  
head(raw_TotalSeqs_df)
#low: 10470-SRR19625128, high: 148864-SRR19625131

# make histogram of raw reads 
raw_TotalSeqs_df %>%
  ggplot(aes(x = TotalSeqs)) + 
  geom_histogram(bins = 50) + 
  scale_x_continuous(limits = c(0, 160000)) + 
  labs(title = "Raw Sequencing Depth Distribution") + 
  theme_classic()

#these did not appear strange during the phylogenetic tree generation, i do not think that i need to remove them

```


## normalize read counts
```{r normalize-read-counts}
#normalization makes comparisons between samples more accurate since different variables can affect read counts
matround <- function(x){trunc(x+0.5)}

scale_reads <- function(physeq, n = min(sample_sums(physeq)), round = "round") {
  
  # transform counts to n
  physeq.scale <- transform_sample_counts(physeq, function(x) {(n * x/sum(x))})
  
  # Pick the rounding functions
  if (round == "floor"){
    otu_table(physeq.scale) <- floor(otu_table(physeq.scale))
  } else if (round == "round"){
    otu_table(physeq.scale) <- round(otu_table(physeq.scale))
  } else if (round == "matround"){
    otu_table(physeq.scale) <- matround(otu_table(physeq.scale))
  }
  
  # Prune taxa and return new phyloseq object
  physeq.scale <- prune_taxa(taxa_sums(physeq.scale) > 1, physeq.scale)
  return(physeq.scale)
}

min(sample_sums(merged_phytree))
#10464
```

## Scale reads and check distribution
```{r scale-reads-and-check-distribution}
scaled_merged_phytree <- 
  merged_phytree %>%
  scale_reads(round = "matround")

# Calculate the read depth 
scaled_TotalSeqs_df <- 
  scaled_merged_phytree %>%
  sample_sums() %>%
  data.frame()
colnames(scaled_TotalSeqs_df)[1] <-"TotalSeqs"

# Inspect
head(scaled_TotalSeqs_df)
max(scaled_TotalSeqs_df) - min(scaled_TotalSeqs_df)

scaled_TotalSeqs_df %>%
  ggplot(aes(x = TotalSeqs)) + 
  geom_histogram(bins = 50) + 
  scale_x_continuous(limits = c(0, 160000)) + 
  labs(title = "Scaled Sequencing Depth at 10470") + 
  theme_classic()

```
samples seem to be at similar range

## PCoA
```{r pcoa}
# Calculate sorensen dissimilarity: Abundance-unweighted of shared taxa
scaled_soren_pcoa <-  
  ordinate(
    physeq = scaled_merged_phytree,
    method = "PCoA",
    distance = "bray", binary = TRUE)

#str(scaled_soren_pcoa)

# Plot the ordination 
soren_horizon_pcoa <- plot_ordination(
  physeq = scaled_merged_phytree,
  ordination = scaled_soren_pcoa,
  color = "horizon",
  shape = "horizon",
  title = "Sorensen Horizon PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = horizon)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = horizon_colors) + 
  theme_bw()
# Show the plot 
soren_horizon_pcoa


#do the same thing for burn year since its working now
soren_burnyear_pcoa <- plot_ordination(
  physeq = scaled_merged_phytree,
  ordination = scaled_soren_pcoa,
  color = "burn_year",
  shape = "horizon",
  title = "Sorensen Burn Year PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = burn_year)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = by_colors) + 
  theme_bw()
# Show the plot 
soren_burnyear_pcoa
# Calculate the BC distance
scaled_BC_pcoa <- 
  ordinate(
    physeq = scaled_merged_phytree,
    method = "PCoA",
    distance = "bray")



# Plot the PCoA for horizons
bray_horizon_pcoa <- 
  plot_ordination(
    physeq = scaled_merged_phytree,
    ordination = scaled_BC_pcoa,
    color = "horizon",
    shape = "horizon",
    title = "Bray-Curtis Horizon PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = horizon)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = c(horizon_colors)) + 
  theme_bw()
bray_horizon_pcoa

#burn year
bray_burnyear_pcoa <- 
  plot_ordination(
    physeq = scaled_merged_phytree,
    ordination = scaled_BC_pcoa,
    color = "burn_year",
    shape = "horizon",
    title = "Bray-Curtis Burn Year PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = burn_year)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = c(by_colors)) + 
  theme_bw()
bray_burnyear_pcoa



unfiltered_unifrac <-
  ordinate(
    physeq=scaled_merged_phytree,
    method="PCoA",
    distance="wunifrac")
#for horizon
w_unifrac <- plot_ordination(
  physeq=scaled_merged_phytree,
  ordination=unfiltered_unifrac,
  color="horizon",
  title= "Weighted Horizon Unifrac PCoA")+
  geom_point(size=4, alpha=0.6, aes(color= horizon))+
  scale_color_manual(values=horizon_colors)
w_unifrac

#for burn year
w_unifrac_burnyear <- plot_ordination(
  physeq=scaled_merged_phytree,
  ordination=unfiltered_unifrac,
  color="burn_year",
  title= "Weighted Horizon Unifrac PCoA")+
  geom_point(size=4, alpha=0.6, aes(color= burn_year))+
  scale_color_manual(values=by_colors)
w_unifrac_burnyear

#combine the plots for horizon
(soren_horizon_pcoa + theme(legend.position = "none")) + 
  (bray_horizon_pcoa + theme(legend.position = "none")) + 
    (w_unifrac + theme(legend.position = "none"))

#combine the plots for burn year
(soren_burnyear_pcoa + theme(legend.position = "none")) + 
  (bray_burnyear_pcoa + theme(legend.position = "none")) + 
    (w_unifrac_burnyear + theme(legend.position = "none"))
```
species found in individual soil horizons seem to cluster together by sorensen
Since I have incorporated both burn year and horizons, it seems that samples cluster together based on the soil horizon.
similar results for the bc; similar clustering based on soil horizon

## NMDS
```{r nmds}
scaled_wUNI_nmds <- 
  ordinate(
    physeq = scaled_merged_phytree,
    method = "NMDS",
    distance = "wunifrac")

#horizon nmds
wUNI_horizon_nmds <- 
  plot_ordination(
    physeq = scaled_merged_phytree,
    ordination = scaled_wUNI_nmds,
    color = "horizon",
    shape = "horizon",
    title = "Weighted Unifrac Horizon NMDS") +
  geom_point(size=5, alpha = 0.5, aes(color = horizon)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = c(horizon_colors)) + 
  theme_bw()
wUNI_horizon_nmds

#burn year nmds
wUNI_burnyear_nmds <- 
  plot_ordination(
    physeq = scaled_merged_phytree,
    ordination = scaled_wUNI_nmds,
    color = "burn_year",
    shape = "burn_year",
    title = "Weighted Unifrac Horizon NMDS") +
  geom_point(size=5, alpha = 0.5, aes(color = burn_year)) +
  scale_shape_manual(values = c(15,16,17,18, 19, 20)) +
  scale_color_manual(values = c(by_colors)) + 
  theme_bw()
wUNI_burnyear_nmds
```
<<<<<<< HEAD
The stress value is low, around 0.05, and is indicative of a good fit for the represented and that horizon may cause the variation observed. NMDS and PCoA plots are also similar, so this is a good sign. 
=======
>>>>>>> 5fb24fe06b44024b1e979f22b3b3ede6b0b8ab55

## PERMANOVA
```{r running-permanova}
# Calculate all three of the distance matrices
scaled_sorensen_dist <- phyloseq::distance(scaled_merged_phytree, method = "bray", binary = TRUE)
scaled_bray_dist <- phyloseq::distance(scaled_merged_phytree, method = "bray")
scaled_wUnifrac_dist <- phyloseq::distance(scaled_merged_phytree, method = "wunifrac")

# make a data frame from the sample_data
# All distance matrices will be the same metadata because they originate from the same phyloseq object. 
metadata_horizon <- data.frame(sample_data(scaled_merged_phytree))

# Adonis test for horizon
adonis2(scaled_sorensen_dist ~ horizon, data = metadata_horizon)

adonis2(scaled_bray_dist ~ horizon, data=metadata_horizon)

adonis2(scaled_wUnifrac_dist ~ horizon, data=metadata_horizon)

metadata_horizon <- data.frame(sample_data(scaled_merged_phytree))

# Adonis test for burn year
metadata_burnyear <- data.frame(sample_data(scaled_merged_phytree))

adonis2(scaled_sorensen_dist ~ burn_year, data = metadata_burnyear)

adonis2(scaled_bray_dist ~ burn_year, data=metadata_burnyear)

adonis2(scaled_wUnifrac_dist ~ burn_year, data=metadata_burnyear)

metadata_burnyear <- data.frame(sample_data(scaled_merged_phytree))
```
<<<<<<< HEAD
Most variation is explained in the weighted unifrac for horizon (almost 0.50 Rvalue), and has the highest F statistic, showing greater differences between samples due to horizon. 
=======
>>>>>>> 5fb24fe06b44024b1e979f22b3b3ede6b0b8ab55

## Running betadispr
```{r betadispr}
#this will tell how sensitive the permanova was to variance in the data
beta_soren_horizon <- betadisper(scaled_sorensen_dist, metadata_horizon$horizon)
permutest(beta_soren_horizon)

#burn year
beta_soren_burnyear <- betadisper(scaled_sorensen_dist, metadata_burnyear$burn_year)
permutest(beta_soren_burnyear)
```

## Phylum Composition
```{r phylum-composition}
# Set the phylum colors

phylum_colors <- c(
  Acidobacteriota = "navy", 
  Actinobacteriota = "darkslategray2", 
  Chloroflexi= "cyan", 
  Crenarchaeota = "firebrick",
"WPS-2" = "magenta",
 Verrucomicrobiota = "#3E9B96",
  Desulfobacterota = "greenyellow",
 Nitrospirota = "yellow",
  Planctomycetota = "#B5D6AA",
  Proteobacteria = "palevioletred1",
  "RCP2-54" = "royalblue",
  Thermoplasmatota = "darkorange", 
  Firmicutes = "olivedrab",
  GAL15 = "greenyellow",
  Gemmatimonadota = "green",
  Myxococcota = "darkorchid1",
 Other = "grey")


```

## Phylum barplots
```{r phylum_barplots}
# get sample data in df
sample_data_df <- data.frame(scaled_merged_phytree@sam_data)

# add reps
modified_sample_data <- sample_data_df %>%
  group_by(burn_year, horizon) %>%
  mutate(replicate_number = row_number())


# fix row names for merge back into phyloseq object
rownames(modified_sample_data) <- modified_sample_data$names

new_sample_data <- sample_data(modified_sample_data)

rownames(new_sample_data) <- modified_sample_data$names

view(new_sample_data)

# merge back into phyloseq

phyloseq_with_reps <- merge_phyloseq(scaled_merged_phytree@otu_table,
                 scaled_merged_phytree@tax_table,
                 scaled_merged_phytree@phy_tree,
                 new_sample_data)
#view(phyloseq_with_reps@sam_data)

phylum_df <- 
  phyloseq_with_reps %>%
  # agglomerate at the phylum level 
  tax_glom(taxrank = "Phylum") %>% 
  # Transform counts to relative abundance 
  transform_sample_counts(function (x) {x/sum(x)}) %>%
  # Melt to a long format 
  psmelt() %>%
  # Filter out Phyla that are <1% - get rid of low abundant Phyla
  dplyr::filter(Abundance > 0.01)

view(phyloseq_with_reps@sam_data)
#I am going to see the unique phyla in the tax df
#phyla <- unique(phylum_df$Phylum)
#view(phyla)


  
# Stacked Bar Plot With All phyla 
# Plot Phylum Abundances - make sure to load phylum_colors 
phylum_df %>%
  # Warning: It's important to have one sample per x value, 
  # otherwise, it will take the sum between multiple samples
<<<<<<< HEAD
  
=======
>>>>>>> 5fb24fe06b44024b1e979f22b3b3ede6b0b8ab55
 #must separate by horizon so that barplots are not stacking on top of each other 
  dplyr::filter(horizon=="A horizon") %>%
  ggplot(aes(x = replicate_number, y = Abundance, fill = Phylum)) + 
  facet_grid(.~burn_year) + 
  geom_bar(stat = "identity", color = "black") + 
  labs(title = "A Horizon Phylum Composition by Burn Year") + 
  scale_fill_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))

phylum_df %>%
  dplyr::filter(horizon=="E horizon") %>%
  ggplot(aes(x = replicate_number, y = Abundance, fill = Phylum)) + 
  facet_grid(.~burn_year) + 
  geom_bar(stat = "identity", color = "black") + 
  labs(title = "E Horizon Phylum Composition by Burn Year") + 
  scale_fill_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))

phylum_df %>%
  dplyr::filter(horizon=="B horizon") %>%
  ggplot(aes(x = replicate_number, y = Abundance, fill = Phylum)) + 
  facet_grid(.~burn_year) + 
  geom_bar(stat = "identity", color = "black") + 
  labs(title = "B Horizon Phylum Composition by Burn Year") + 
  scale_fill_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))

#make each phylum its own row
#a horizon
phylum_df %>%
  dplyr::filter(horizon=="A horizon") %>%
  ggplot(aes(x = replicate_number, y = Abundance, fill = Phylum)) + 
  facet_grid(Phylum~burn_year) + 
  geom_bar(stat = "identity", color = "black") + 
  labs(title = "A Horizon Phylum Composition by Burn Year") + 
  scale_fill_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1), strip.text.y.right = element_text(angle = 0))

#e horizon
phylum_df %>%
  dplyr::filter(horizon=="E horizon") %>%
  ggplot(aes(x = replicate_number, y = Abundance, fill = Phylum)) + 
  facet_grid(Phylum~burn_year) + 
  geom_bar(stat = "identity", color = "black") + 
  labs(title = "E Horizon Phylum Composition by Burn Year") + 
  scale_fill_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1), strip.text.y.right = element_text(angle = 0))

#b horizon
phylum_df %>%
  dplyr::filter(horizon=="B horizon") %>%
  ggplot(aes(x = replicate_number, y = Abundance, fill = Phylum)) + 
  facet_grid(Phylum~burn_year) + 
  geom_bar(stat = "identity", color = "black") + 
  labs(title = "B Horizon Phylum Composition by Burn Year") + 
  scale_fill_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1), strip.text.y.right = element_text(angle = 0))

#look at a specific phylum
phylum_df %>%
  dplyr::filter(Phylum == "Acidobacteriota") %>%
  # build the plot 
  ggplot(aes(x = horizon, y = Abundance, 
             fill = horizon, color = horizon)) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Acidobacteriota Phylum Abundance") + 
  #scale_color_manual(values = by_colors) + 
  #scale_fill_manual(values = by_colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")

#look at a specific phylum
phylum_df %>%
  dplyr::filter(Phylum == "Acidobacteriota") %>%
  # build the plot 
  ggplot(aes(x = burn_year, y = Abundance, 
             fill = burn_year, color = burn_year)) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Acidobacteriota Phylum Abundance") + 
  #scale_color_manual(values = by_colors) + 
  #scale_fill_manual(values = by_colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")

```
<<<<<<< HEAD
Acidobacteriodota seems to have a higher composition at lower horizons, but I can only tell this from the bar plots.  These bacteria are typically found in hot springs so it is interesting seeing it appear. 
=======
Acidobacteriodota seems to be higher at the lower horizons!
It also seems higher on more burned soils.
>>>>>>> 5fb24fe06b44024b1e979f22b3b3ede6b0b8ab55

# Do the same thing for family and genus
```{r family-and-genus}
family_df <- 
  phyloseq_with_reps %>%
  # agglomerate at the phylum level 
  tax_glom(taxrank = "Family") %>% 
  # Transform counts to relative abundance 
  transform_sample_counts(function (x) {x/sum(x)}) %>%
  # Melt to a long format 
  psmelt() %>%
  # Filter out Phyla that are <1% - get rid of low abundant Phyla
  dplyr::filter(Abundance > 0.01)

view(phyloseq_with_reps@sam_data)


family_df %>%
 #must separate by horizon so that barplots are not stacking on top of each other 
  dplyr::filter(horizon=="A horizon") %>%
  dplyr::filter(Family == "Acidothermaceae") %>%
 ggplot(aes(x = burn_year, y = Abundance, 
             fill = burn_year, color = burn_year)) + 
  facet_wrap(.~Family, scales = "fixed", nrow = 1) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Acidothermaceae Family Abundance") + 
  scale_color_manual(values = by_colors) + 
  scale_fill_manual(values = by_colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "bottom")

family_df %>%
 #must separate by horizon so that barplots are not stacking on top of each other 
  dplyr::filter(horizon=="E horizon") %>%
  dplyr::filter(Family == "Acidothermaceae") %>%
 ggplot(aes(x = burn_year, y = Abundance, 
             fill = burn_year, color = burn_year)) + 
  facet_wrap(.~Family, scales = "fixed", nrow = 1) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Acidothermaceae Family Abundance E horizon") + 
  scale_color_manual(values = by_colors) + 
  scale_fill_manual(values = by_colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "bottom")

family_df %>%
 #must separate by horizon so that barplots are not stacking on top of each other 
  dplyr::filter(horizon=="B horizon") %>%
  dplyr::filter(Family == "Acidothermaceae") %>%
 ggplot(aes(x = burn_year, y = Abundance, 
             fill = burn_year, color = burn_year)) + 
  facet_wrap(.~Family, scales = "fixed", nrow = 1) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Acidothermaceae Family Abundance B horizon") + 
  scale_color_manual(values = by_colors) + 
  scale_fill_manual(values = by_colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "bottom")
```

Acidothermacaea decreases at the E horizon (scales changed), which I thought would be the opposite due to the previous graph. Abundance decreased further in the B horizon.

```{r}
genus_df <- 
  phyloseq_with_reps %>%
  # agglomerate at the phylum level 
  tax_glom(taxrank = "Genus") %>% 
  # Transform counts to relative abundance 
  transform_sample_counts(function (x) {x/sum(x)}) %>%
  # Melt to a long format 
  psmelt() %>%
  # Filter out Phyla that are <1% - get rid of low abundant Phyla
  dplyr::filter(Abundance > 0.01)

view(phyloseq_with_reps@sam_data)


genus_df %>%
 #must separate by horizon so that barplots are not stacking on top of each other 
  dplyr::filter(horizon=="A horizon") %>%
  dplyr::filter(Genus == "Acidothermus") %>%
 ggplot(aes(x = burn_year, y = Abundance, 
             fill = burn_year, color = burn_year)) + 
  facet_wrap(.~Family, scales = "fixed", nrow = 1) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Acidothermus Family A horizon Abundance") + 
 scale_color_manual(values = by_colors) + 
  scale_fill_manual(values = by_colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "bottom")

genus_df %>%
 #must separate by horizon so that barplots are not stacking on top of each other 
  dplyr::filter(horizon=="E horizon") %>%
  dplyr::filter(Genus == "Acidothermus") %>%
 ggplot(aes(x = burn_year, y = Abundance, 
             fill = burn_year, color = burn_year)) + 
  facet_wrap(.~Family, scales = "fixed", nrow = 1) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Acidothermus Family E horizon Abundance") + 
  scale_color_manual(values = by_colors) + 
  scale_fill_manual(values = by_colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "bottom")

genus_df %>%
 #must separate by horizon so that barplots are not stacking on top of each other 
  dplyr::filter(horizon=="B horizon") %>%
  dplyr::filter(Genus == "Acidothermus") %>%
 ggplot(aes(x = burn_year, y = Abundance, 
             fill = burn_year, color = burn_year)) + 
  facet_wrap(.~Family, scales = "fixed", nrow = 1) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Acidothermus Family B horizon Abundance") + 
  scale_color_manual(values = by_colors) + 
  scale_fill_manual(values = by_colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "bottom")
```
Abundance for acidothermus is low at A horizon. I am wondering if it increases with depth? It actually decreases. Abundance of higher taxonomic levels seems to decrease with depth. 
However, each of the investigated rank for the phylum/family/genus increased with burned soil year. From this, It would appear that more burned soils faciliated an environment where these taxa would have higher composition. 

# Session Information
```{r session-information}
devtools::session_info()
```

