---
title: "05_communityanalysis"
author: "Ri Desroches"
date: "`r Sys.Date()`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "../figures/05_communityanalysis/")

```

Note before document start:

I was originally going to filter out the sequences that were unclassified at the kingdom level and run the community analyses on both the filtered and unfiltered datasets. However, I ended up not including both in the final analysis for time reasons. The start of the code may reflect that originally I planned to complete the analyses for both, but this was not actually feasible for me to do within the time constraints of the semester. Because of this, community composition analyses are run on the unfiltered dataset, which does include all of the reads that are unclassified at the Kingdom level.

# Setup 

## set seed
```{r set-seed}

set.seed(0110)

```

## Load Packages

```{r load-packages}
pacman::p_load(tidyverse, devtools, phyloseq, patchwork, vegan,
               install = FALSE)
```


## Load data

```{r load-data}
load("data/04_phylogeny/phytree_preprocessed_physeq.RData")

#set color schemes as well

#color vectors based on burn time frame
timeframe_colors <- c(
  "1"="darkorange1",
  "2"="goldenrod2",
  "4"="olivedrab",
  "unburned"="darkgreen")


#color vectors based on sample horizon
sample_horizon_colors <- c(
  "A horizon"="rosybrown1",
  "B horizon"="lightsalmon3",
  "E horizon"= "darkred")

# Color vectors for horizon+burn timeframe combined PCoA
combined_horizon_timeframe <- c(
  "lightgoldenrod",
  "orange",
  "orangered2",
  "orangered4",
  "lightcyan",
  "darkslategray2",
   "deepskyblue3",
   "dodgerblue4",
  "yellow",
  "olivedrab2",
  "olivedrab",
  "forestgreen")

```


## Goals :

1. Evaluate sequencing depth
2. Remove low-depth samples
3. Normalize read counts
4. Calculate community dissimilarities with three metrics:
  - Sorensen
  - Bray-Curtis
  - Abundance weighted unifrac
5. Visualize the community data with:
  - PCoA
  - NMDS
6. Stats analyses


## Explore read counts

```{r explore-read-counts}
raw_totalseqs_df <-
  unfiltered_rooted_tree_physeq %>%
  sample_sums() %>%
  data.frame()

colnames(raw_totalseqs_df)[1] <- "TotalSeqs"

head(raw_totalseqs_df)


# make histogram

raw_totalseqs_df %>%
  ggplot(aes(x = TotalSeqs)) + 
  geom_histogram()+ 
  labs(title = "Raw Sequencing Depth Distribution")
```

The lowest seq depth sample is 14380 seqs

Keep all samples, there is a very wide range of seqs

The only outlier is the one with 200k+ sequences, see if it stands out during community analyses

That sample is : SRR19625131

```{r SRR19625131-sample-info}
sample_data <- unfiltered_rooted_tree_physeq@sam_data %>%
  tibble::as_tibble()

dplyr::filter(sample_data, names == "SRR19625131")

```

That sample is an A horizon, 1-year burned forest

I also see that I have to separate burn year into its own metadata column again, but wrote the code previously and it didn't save it as an object. Oops!

Fixing before moving on

## Fixing metadata.. again

```{r adding-burn-year-metadata}
#Used code from 03_biodiversity.Rmd


#Information to add:
  # burn status (1 year, 2 year, 4 year, unburned)
  #separate from "sample site"
  # goal is to have a column that just has timeframe of burn without A-F sampling information


#extract burn time frame information from one of the 4 objects
unrooted_physeq@sam_data$Site

burn_timeframe <- gsub("^CH", "unburned", gsub("[A-Z]$", "", unrooted_physeq@sam_data$Site))

# fix unrooted physeq

unrooted_physeq@sam_data$Site <- burn_timeframe

sample_data(unrooted_physeq)$burn_timeframe <-
  sample_data(unrooted_physeq)$Site

# fix midroot physeq

unfiltered_rooted_tree_physeq@sam_data$Site <- burn_timeframe

sample_data(unfiltered_rooted_tree_physeq)$burn_timeframe <- 
  sample_data(unfiltered_rooted_tree_physeq)$Site

# fix unrooted filtered physeq

unrooted_physeq_filtered@sam_data$Site <- burn_timeframe

sample_data(unrooted_physeq_filtered)$burn_timeframe <-
  sample_data(unrooted_physeq_filtered)$Site

# fix rooted filtered physeq

filtered_rooted_tree_physeq@sam_data$Site <- burn_timeframe

sample_data(filtered_rooted_tree_physeq)$burn_timeframe <-
  sample_data(filtered_rooted_tree_physeq)$Site

# combined metadata groups for colors
combined_groups <- interaction(sample_data(filtered_rooted_tree_physeq)$burn_timeframe, sample_data(filtered_rooted_tree_physeq)$horizon)

combined_groups


```

## Normalize read counts - import function

```{r normalize-read-counts-function}

# Function to scale reads: http://deneflab.github.io/MicrobeMiseq/ 
# Scales reads by 
# 1) taking proportions
# 2) multiplying by a given library size of n
# 3) rounding 
# Default for n is the minimum sample size in your library
# Default for round is floor

matround <- function(x){trunc(x+0.5)}

scale_reads <- function(physeq, n = min(sample_sums(physeq)), round = "round") {
  
  # transform counts to n
  physeq.scale <- transform_sample_counts(physeq, function(x) {(n * x/sum(x))})
  
  # Pick the rounding functions
  if (round == "floor"){
    otu_table(physeq.scale) <- floor(otu_table(physeq.scale))
  } else if (round == "round"){
    otu_table(physeq.scale) <- round(otu_table(physeq.scale))
  } else if (round == "matround"){
    otu_table(physeq.scale) <- matround(otu_table(physeq.scale))
  }
  
  # Prune taxa and return new phyloseq object
  physeq.scale <- prune_taxa(taxa_sums(physeq.scale) > 0, physeq.scale)
  return(physeq.scale)
}

```

## Normalize read counts 

```{r normalize-read-counts}
min(sample_sums(unfiltered_rooted_tree_physeq))

# scale reads to 14380

scaled_unfiltered_rooted_physeq <-
  unfiltered_rooted_tree_physeq %>%
  scale_reads(round="matround")


scaled_totalseqs_df <-
  scaled_unfiltered_rooted_physeq %>%
  sample_sums() %>%
  data.frame() 

colnames(scaled_totalseqs_df) <- "seq_counts"

head(scaled_totalseqs_df)

max(scaled_totalseqs_df) - min(scaled_totalseqs_df)


# save normalized reads
save(scaled_unfiltered_rooted_physeq, file="data/05_communityanalysis/normalizedreads.RData")

#Plot as histogram

scaled_totalseqs_df %>%
  ggplot(aes(x=seq_counts)) +
  geom_histogram()

```

Looks good, they are all in a similar range.

## Community analyses

### Sorensen

```{r Sorensen}

unfiltered_sorensen <-
  ordinate(
    physeq=scaled_unfiltered_rooted_physeq,
    method="PCoA",
    distance="bray", binary=TRUE)

SORENSENPLOT <-
  plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_sorensen,
  color=combined_groups,
  title="Sorenseen PCoA of Soil Horizon and Burn Timeframe")+
  geom_point(size=4, alpha=0.6, aes(color=combined_groups))+
  scale_color_manual(values=combined_horizon_timeframe)

SORENSENPLOT


```

### Bray-Curtis

```{r bray-curtis}
unfiltered_bray <-
  ordinate(
    physeq=scaled_unfiltered_rooted_physeq,
    method="PCoA",
    distance="bray", binary=FALSE)


 BRAYPLOT <-  plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_bray,
  color=combined_groups,
  title="Bray Curtis PCoA of Soil Horizon and Burn Timeframe")+
  geom_point(size=4, alpha=0.6, aes(color=combined_groups))+
  scale_color_manual(values=combined_horizon_timeframe) 
 
BRAYPLOT

```

### Weighted Unifrac

```{r weighted-unifrac}

unfiltered_unifrac <-
  ordinate(
    physeq=scaled_unfiltered_rooted_physeq,
    method="PCoA",
    distance="wunifrac")

WUNIFRAC <-
  plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_unifrac,
  color=combined_groups,
  title="Weighted Unifrac PCoA of Soil Horizon and Burn Timeframe")+
  geom_point(size=4, alpha=0.6, aes(color=combined_groups))+
  scale_color_manual(values=combined_horizon_timeframe)

WUNIFRAC


## Another way of viewing, with maybe a better legend


plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_unifrac,
  color="horizon",
  title="Unfiltered-Kingdom Dataset Weighted Unifrac PCoA")+
  geom_point(size=4, alpha=0.6, aes(color=horizon, shape=burn_timeframe))+
  scale_color_manual(values=sample_horizon_colors)



```

All three plots show a clear separation of samples based on soil horizon, but not burn timeframe. Even within each soil horizon, there is no clear separation based on burn timeframe. This is not at all what I expected! I thought that they would cluster based on both soil horizon AND burn timeframe, but this indicates to me that the burns have minimal impact on the community dissimilarities.

This is especially surprising to me because of the biodiversity boxplots in 03_biodiversity and the crazy tree from 04_phylogeny anaylsis. I thought that maybe lowly abundant / rare taxa across burn timeframes might cause some community differences, but even when looking at Sorensen dissimilarity which deals with just presence/absence, there still isn't a large difference (at least when plotting.) This will be examined further with the permanova/adonis calculations.

###NMDS

```{r NMDS}
unfiltered_1_nmds <-
    ordinate(
    physeq=scaled_unfiltered_rooted_physeq,
    method="NMDS",
    distance="bray", binary=TRUE)

unfiltered_2_nmds <-
  ordinate(
    physeq=scaled_unfiltered_rooted_physeq,
    method="NMDS",
    distance="bray", binary=FALSE)

unfiltered_3_nmds <-
  ordinate(
    physeq=scaled_unfiltered_rooted_physeq,
    method="NMDS",
    distance="wunifrac")
  

plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_1_nmds,
  color=combined_groups,
  title="Sorensen NMDS of Soil Horizon and Burn Timeframe")+
  geom_point(size=4, alpha=0.6, aes(color=combined_groups))+
  scale_color_manual(values=combined_horizon_timeframe)

plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_2_nmds,
  color=combined_groups,
  title="Bray-Curtis NMDS of Soil Horizon and Burn Timeframe")+
  geom_point(size=4, alpha=0.6, aes(color=combined_groups))+
  scale_color_manual(values=combined_horizon_timeframe)

plot_ordination(
  physeq=scaled_unfiltered_rooted_physeq,
  ordination=unfiltered_3_nmds,
  color=combined_groups,
  title="Weighted Unifrac NMDS of Soil Horizon and Burn Timeframe")+
  geom_point(size=4, alpha=0.6, aes(color=combined_groups))+
  scale_color_manual(values=combined_horizon_timeframe)

```


The pattern stays consistent with NMDS - even when "smushing" the multiple axes of diversity, the soils do not separate based on burn timeframe. It still seems that the primary differences in community composition comes from the soil horizon.

It is interesting that in almost all of the PCoAs and NMDS graphs show the E horizon as an intermediate between the A and B soil horizons. I thought this was strange, since I initially thought that the E horizon was the deepest soil layer. But this is not true! The E horizon sits between the A and B horizons, so it makes sense that the E horizon communities seem like an intermediate between the A and B horizons.



### Permanova - Burn timeframe

```{r permanova-burn}

scaled_sorensen <- phyloseq::distance(scaled_unfiltered_rooted_physeq, method = "bray", binary = TRUE)
scaled_bray <- phyloseq::distance(scaled_unfiltered_rooted_physeq, method = "bray")
scaled_unifrac <- phyloseq::distance(scaled_unfiltered_rooted_physeq, method = "wunifrac")

metadata <- data.frame(sample_data(scaled_unfiltered_rooted_physeq))

adonis2(scaled_sorensen ~ burn_timeframe, data=metadata)

adonis2(scaled_bray ~ burn_timeframe, data=metadata)

adonis2(scaled_unifrac ~ burn_timeframe, data=metadata)

```


### Permanova - Horizon 
```{r permanova-horizon}

adonis2(scaled_sorensen ~ horizon, data=metadata)

adonis2(scaled_bray ~ horizon, data=metadata)

adonis2(scaled_unifrac ~ horizon, data=metadata)

```

The community dissimilarities based on soil horizon are statistically significant, but the community dissimilarities based on burn timeframe are not. I wonder if this is still true if we look within horizon groups?

I would have liked to look at this more deeply to answer that question, but wasn't sure how to filter out the facets to examine without disrupting the tree portion of the physeq object. 